keps/sig-api-machinery/3610-env-injection/README.md

```diff
@@ -0,0 +1,855 @@
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP defines a new annotation `glabel.env` and introduces a mechanism that can override current environment variables on pods.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+This seems to be a common requirement.
```

Comment:
I would like to see a much more fleshed out motivation for this. Specifically, I don't think this has adequately answered the question of why this should be a built in controller, and not a webhook (or eventually [CEL admission policy](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control))

Reply:
Do you mean another proposal like below?
- User can use CEL admission fro a new CRD
- And using a webhook to add env according to the new CRD

Reply:
We can use CEL admission policy to validate if an env value is meeting the requirement.

https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/



object.envars.filter(e, e.name == 'MY_ENV').all(e, e.value.matches('^[a-zA-Z]*) | Validate the 'value' field of a listMap entry where key field 'name' is 'MY_ENV'
-- | --

Reply:
It's not clear to me how this is different in nature from the [PodPreset](https://github.com/kubernetes/kubernetes/pull/94090) feature that never progressed past alpha and was removed.

Using admission webhooks for now and eventually something like mutating CEL admission to do this injection seems better at first glance than crafting a built-in API with this narrow of a use case.

Reply:
> Using admission webhooks for now and **eventually something like mutating CEL admission** to do this injection seems better at first glance than crafting a built-in API with this narrow of a use case.

@liggit thanks for your information here and at sig-auth meeting.
I have searched https://github.com/kubernetes/enhancements/issues/2876 and it is in the future plan part.

> CEL might be used in Kubernetes for extensibility beyond CRD validation. The future plans section of this KEP explains how CEL might be used for general admission control, defaulting and conversion. This KEP aims to prove the utility of CEL for both the immediate use case (CRD validation) and these future use cases. This KEP also aims to use CEL in a way that is congruent with these future use cases.

Is there any process for mutating CEL admission? The use case here is one of the defaulting part IMO.

/close

Reply:
> Is there any process for mutating CEL admission? The use case here is one of the defaulting parts IMO.

Not yet, but it's a goal... use/integration of CEL is being worked on incrementally to help us gain more experience with integration CEL, with the following items progressing/completing ahead of mutating capabilities:

* CEL validation of CRDs (https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language) went to beta in 1.25 and is actively progressing toward GA
  > Future Work
  > CEL for General Admission Control
  > The ability to use CEL to extend Kubernetes admission control could allow for both validating and mutating admission control without the use of webhooks.
* CEL validating admission (https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3488-cel-admission-control), introduced in alpha in 1.26 and actively being worked on in 1.27
  > Mutation support. While we believe this enhancement could be extended in the future to support mutations, we believe it is best handled as a separate enhancement. That said, we need to keep mutation in mind in this KEP to ensure we design it in such a way that we don't obviously paint ourselves into a corner where mutation would be difficult to introduce later.

cc @jpbetz @cici37 @tallclair